pandas/_libs/algos.pyx

```diff
@@ -1158,34 +1157,14 @@ def rank_2d(
 
     n, k = (<object>values).shape
     ranks = np.empty((n, k), dtype='f8')
-    # For compatibility when calling rank_1d
-    labels = np.zeros(k, dtype=np.int64)
 
-    if rank_t is object:
-        try:
-            _as = values.argsort(1)
-        except TypeError:
-            values = in_arr
```

Comment:
This is the source of the bug - if `axis=0`, values is supposed to be transposed (as done earlier in the function) - setting equal to `in_arr` makes `values` the wrong shape since `in_arr` was not transposed